Generic parsing restructuring

mfitzp opened the pull request:
This is a zero side-effect restructuring of clinto to prepare for
other parsers. Parsers are defined off a base class which provides
a series of stub functions that must be overridden to provide
a new parser.  The initialisation of a parser accepts a script
path and source code which is then tested, first for file extension
then for the file containing a magic token (e.g. ArgumentParser) that
identifies that file as likely to be using a particular parser.

If these match the parsing is performed and the parse object parsed(!)

If this final step is sucessful parsing will halt, otherwise it will
continue on to try the next parser.

Individual parsers are isolated in separate files (e.g. argparse.py)
and are free to use their own logic and internals to generate the output.
The json output of sucessfully processed parsers is available via the
public parser.Parser class (and is the get_script_description for debug).

This incorporates the changes for exception monkey patching.

mfitzp committed:
Generic parsing restructuring

This is a zero side-effect restructuring of clinto to prepare for
other parsers. Parsers are defined off a base class which provides
a series of stub functions that must be overridden to provide
a new parser.  The initialisation of a parser accepts a script
path and source code which is then tested, first for file extension
then for the file containing a magic token (e.g. ArgumentParser) that
identifies that file as likely to be using a particular parser.

If these match the parsing is performed and the parse object parsed(!)

If this final step is sucessful parsing will halt, otherwise it will
continue on to try the next parser.

Individual parsers are isolated in separate files (e.g. argparse.py)
and are free to use their own logic and internals to generate the output.
The json output of sucessfully processed parsers is available via the
public parser.Parser class (and is the get_script_description for debug).

This incorporates the changes for exception monkey patching.

mfitzp committed:
Fix tests

mfitzp commented:
@Chris7 this is an alternative go at generic parsing. It doesn't really _do_ anything other than restructure the package to allow us to drop in new parsers (e.g. for click, etc.) in a isolated way.

Chris7 commented on clinto/parsers/base.py:
```diff
@@ -0,0 +1,63 @@
+    def check_valid(self):
+        if self.script_path:
+            ext = os.path.splitext(os.path.basename(self.script_path))[1]
+            return ext in self.extensions and self.contains in self.script_source
```
I don't think the 'contains'/extension will work. For contains, you can subclass custom parsers and call them whatever. For the extension, an example is wooify, which doesn't use a .py extension (which you cannot use for windows compatibility)

mfitzp replied:
Ah crap, good point. That is a bit of an issue :/ ...we need a way to avoid running every parser on every script (particularly if we start supporting other languages, e.g. R where we're going to have to fire up an interpreter). Any ideas? 

The contains is true but I was assuming that to subclass you'd still need the name of it in the file (since we're only allowing standalone files).

Chris7 replied:
For contains, an example I do in my own code is to define an argument parser in a common file that all my scripts inherit from, so argparse actually never enters my text (but you can figure out it's there by doing issubclass(obj, ArgumentParser).

It's something like:
from . import CustomParser
p = CustomParser(...)
p.add_tsv_file()
p.add_fasta_file()

etc. That lets me have a consistent interface for all my scripts.

mfitzp replied:
How are you importing from the common file in the standalone script? Is it installed as a module?

The problem is though we can test for a ArgumentParser we can only do that _after_ we have the ArgumentParser. The check was to see how likely it would be to get a parser _before_ getting it. 

Otherwise we'll have repeatedly process the script (for argparse, click, etc. etc.) until we hit a match. I was hoping to shortcut that... might not be possible.

How about rather than a blunt check it returns a heuristic 'score' for each parser, and then we can test them in order? That will cut down on processing for most things but allow it to work regardless.

Chris7 replied:
The import is coming from the package, it's a module. from . import CustomParser should be from mymodule.templates import CustomParser.

I did it by loading the file and then making diagnostics, like this:

+SUPPORTED_PARSERS = {
-    'argparse': {
-        'methods': ['**package**', 'ArgumentParser'],
-        'parser': argparse_parser.ArgparseParser,
-    },
-    'click': {
-        'methods': ['**package**', 'parser', 'ClickException'],
-        'parser': click_parser#.ClickParser,
-    },
  +}

That would go down the list of objects founds, and try to determine which parser is active via classes specific to that parser. That was the heuristic I was using. Though that obviously breaks sometimes from people putting parse_args where it shouldn't be.

Ultimately, I think we're going to have to experiment with AST to really get this right in all cases...

mfitzp replied:
But to test which parser you have you're going to have to get the parser first. That means for a click script you have to first try and process it as argparse - and that has to happen for every script. It's not so bad now but once we've implemented other command parsers (and languages) it's going to be a lot of wasted processing on those scripts.

I'm more looking for a way to prefer one parser over another (e.g. try and locate ClickParser first for scripts that _look like_ click scripts). I'll put together an implementation of the heuristic thingy to demonstrate it .... for example it could be as simple as "does the file contain 'python' or 'import' or 'lambda' ...then it's more likely to be a Python file than R, test the Python loaders first".

I'm getting convinced that we can sidestep AST altogether by monkey patching and throwing exceptions - it works without problems on every argparse + click script I've tested. The advantage of doing it this way around is that once you have the object you _know_ what it is, because otherwise you wouldn't have it. The only way you can get a result from monkey-patching argparse is if the parser is argparse! :)

mfitzp committed:
First implementation of docopt support using custom parser.

This works with the tested demo scripts. Doesn't handle subparsers
or anything else funky like that.

mfitzp committed:
Implement heuristic ranking for parsers

Replacing the previous fixed requirement for extension and
file contents is a ranking approach which scores scripts against
a list of common contents, patterns and extensions. The result is
a number 0..1 reflecting how closely the script matches the
pattern. For most scripts this will result in the closest match
of parser being applied first to speed up parsing.

Possible improvements would be a base ranking value
for each parser to prefer the most common one when
two scripts score the same (e.g. always prefer argparse first).

mfitzp committed:
Fix tests for new import argparse_ style due to module clash

mfitzp committed:
Fix init for script_path=None

mfitzp commented:
This implements the scoring heuristic thing that ranks the parsers before applying them based on definable characteristics. All your existing scripts should still work, and assuming they contain any of the following they should hit the argparse parser first. 

``` python
    def heuristic(self):
        return [
            self.script_ext in ['.py', '.py3', '.py2'],
            'argparse' in self.script_source,
            'ArgumentParser' in self.script_source,
            '.parse_args' in self.script_source,
            '.add_argument' in self.script_source,
        ]
```

The score is calculated as a ratio of matches so 2 matches = 2/5 = 0.4. This is compared relative to the scores for other parsers, e.g. for docopt:

``` python
    def heuristic(self):
        return [
            self.script_ext in ['.py', '.py3', '.py2'],
            'docopt' in self.script_source,
            '__doc__' in self.script_source,
        ]
```

...so here 2 matches would score 0.6. This has no effect on _whether_ a parser is run just the order.

I re-added the script_name parameter to keep it compatible with the current version of Wooey, but it's not required for extracting the parsers — passing in '**main**' as the script name allows the `if '__name__' == '__main__':` blocks used in a lot of scripts to work.

mfitzp committed:
Add error passthrough.

mfitzp committed:
Exec not execfile

Chris7 commented on clinto/parser.py:
```diff
@@ -1,233 +1,47 @@
-        return json.dumps(self.get_script_description())
+    def error(self):
+        if self.parser:
+            return self.parser.error
```
Might want to make this getattr(self.parser, 'error', '') and the default return to '', since error is meant to be a string.

mfitzp replied:
I left it as `self.parser.error` since the `error` attribute is set in the init for the base class (it'll always be set). But I did change the default return. Also noticed that `get_script_description` stub was wrongly returning a `None` rather than empty `dict` for the 'json' so I've changed that to `{}`

mfitzp committed:
Fix default return values for parser interface and BaseParser
